### ISSUE TITLE
[1.4] [Proposal] NewProjectile owner parameter change

### ISSUE THREAD

**Issue by direwolf420** [bluemagic123/tModLoader#2181]
### Do you intend to personally contribute/program this feature?

No

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

Currently, `Projectile.NewProjectile` and similar all have the optional `int Owner = 255` parameter. There are multiple issues with it when you do not specify it yourself:
1. Projectile deals no damage when it clearly has damage, friendly status, etc.
2. Projectile disappears on tile collision, when it (for example) should've bounced instead
3. Projectile may not appear for other clients in multiplayer
4. Many more that I don't remember at the time of writing this

This is most often observed in new modders that are not familiar with how the mechanic of spawning projectiles work (and can you blame them?), more on that below.

**Proposal**: To avoid these issues, I suggest two approaches:
1. Make it non-optional:
    * This forces the modder to think about what to put as the proper parameter (`Main.myPlayer`)
    * _This (as of 1.4.3.6) only causes a single error in an NPCs AI spawning a projectile which can be patched by tml, or might be fixed in future versions of Terraria (I notified the devs)_

2. Remove it entirely and/or assign `Main.myPlayer` to it automatically
    * _Implementation: Internalize existing ones and replace by tml methods without the parameter (might cause vanilla side effects if it fails surrounding checks, more on that below)_
    * Avoids boilerplate assignment
    * If it's removed, it causes porting/adapting overhead since it's 1 parameter less
    * If it's made redundant (ignoring parameter value basically), no overhead
    
**Technical explanation:**
Various logic deciding the "fate" of the projectile is only ran by the owner, such as dealing damage, or despawning from timeLeft running out or from colliding with tiles.
Projectiles should only ever be spawned from **one** source, either a single client, or the server, to prevent duplicate projectiles. This is facilitated through surrounding code, such as:
 ```c#
if (Main.myPlayer == Player.whoAmI) //or if (Main.myPlayer == Projectile.owner)
{
    //Owner can only be Main.myPlayer here
    Projectile.NewProjectile(...);
}
```
or
```c#
if (Main.netMode != NetmodeID.MultiplayerClient)
{
    //Owner can only be Main.myPlayer here aswell for proper functionality
    Projectile.NewProjectile(...);
}
```

In multiplayer, failure to do so while using `Main.myPlayer` as the owner causes duplicate projectiles which are all "valid" given the context. If you were to use `Player.whoAmI` in a client context, then duplicate projectiles would not occur as they would fail to properly sync (sync check is behind the exact `Owner == Main.myPlayer` check), ironically acting as a "failsafe" for not having the owner properly assigned.
Defaulting to `Main.myPlayer` as the owner as approach 2. would then only cause this particular issue in multiplayer, and fix the other ones mentioned at the start, essentially trading a whole bucket of issues for a singular issue (both caused by an oversight or for not handling spawning properly).

tl;dr: Reduce number of issues modders have to face (and possibly spend alot of time figuring out the solution for) when omitting the owner parameter for `Projectile.NewProjectile`.

### What does this proposal attempt to solve or improve?

above

### Which (other) solutions should be considered?

above

**Comment by JavidPack** [comment]
Was there any consensus on Terraria looking into this? @Chicken-Bones

**Comment by direwolf420** [comment]
![Screenshot_6](https://user-images.githubusercontent.com/15894498/190498798-27465a85-91f5-402c-a721-f3352f0d0250.png)

Yes
